lib/class-perfecty-push-lib-cron-check.php

```diff
@@ -0,0 +1,217 @@
+<?php
+
+use Perfecty_Push_Lib_Log as Log;
+
+/***
+ * Check cron functionality
+ */
+class Perfecty_Push_Lib_Cron_Check {
+
+	private const CRON_HOOK = 'perfecty_push_cron_check';
+
+	private const SCHEDULE_OFFSET = 1800; // 30 minutes
+
+	private const SUCCESS_RUN_LOG_MSG = 'Cron monitor runned';
+
+	private const NEXT_SCHED_LOG_MSG = 'Next cron check scheduled at: ';
+
+	private const NEVER_SCHED_LOG_MSG = 'Cron monitor was not scheduled';
+
+	private const NOT_RUNNED_LOG_MSG = 'Cron monitor was scheduled but not executed';
+
+	private const LIMIT_LOGS_FOR_CALC = 10;
+
+	/**
+	 * The cron running status.
+	 *
+	 * @since    1.2.0
+	 * @access   private
+	 * @var      bool    $cron_working    The cron running status.
+	 */
+	private static $cron_working = false;
+
+	/**
+	 * Initialize the class.
+	 *
+	 * @since    1.2.0
+	 */
+	public function __construct() {
+		$this->schedule_cron_job();
+	}
+
+	/**
+	 * Schedules the cron job once (this is the callback of the cronjob too).
+	 */
+	public function schedule_cron_job() {
+		if ( ! wp_next_scheduled( self::CRON_HOOK ) ) {
+			$now                 = time();
+			$next_scheduled_time = $now + self::SCHEDULE_OFFSET;
+			wp_schedule_single_event( $next_scheduled_time, self::CRON_HOOK );
+			Log::info(
+				self::SUCCESS_RUN_LOG_MSG . ' - ' . self::NEXT_SCHED_LOG_MSG . \
+				get_date_from_gmt( gmdate( 'Y-m-d H:i:s', intval( $next_scheduled_time ) ) ),
+				'Y-m-d H:i:s'
+			);
+		}
+	}
+
+	/**
+	 * Checks if next scheduled action is in the future and returns cron status
+	 *
+	 * @return bool
+	 */
+	public static function get_cron_status() {
+		$now = time();
+		/**
+		 * When wp-cron is disabled and a system cron is running we don't want to annoy users
```

Comment:
Hi @rwngallego, not sure about what you said.
I think that what is problematic for the plugin is to set DISABLE_WP_CRON = true.
The plugin executes the "first" batch when wp-cron.php is called (by system cron, or by a page request) and the scheduled time is <= now.
Then, next batches are "scheduled" to be called "now" (that is, immediately when the next call to wp-cron.php is done).
These further calls are triggered in the same way: usually, when there is no "DISABLE_WP_CRON = true" in wp-config.php, it is immediately because wp-cron.php is called just when a user opens a page (maybe, following the notification received in the first batch).
However, this is not what will happen if DISABLE_WP_CRON = true is set, because in this case the next call to wp-cron.php will be received only when the system cron will run.
So, if the system cron is set to run every 10 minutes, you will have a delay of 10 minutes between each batch of the same notification.
This is a general "issue" and it is completely unrelated to the cron-check or the log lib.
Substantially, to have "immediate" execution of batches, there should be no DISABLE_WP_CRON = true and the site has to have "traffic"
We can write this somewhere, but I think that if a wpadmin sets DISABLE_WP_CRON = true he should know what it means; maybe he doesn't but we cannot modify the cron behaviour or the wp-cron system.

For what I know, to have better performance on Time To First Byte measurement, many admins disable wp-cron in their config.
The calculations I made in the code are just to chose when we show an alert to say "something is wrong with cron" and it is made to avoid showing the alert when an admin (consciously) decided to disable wp-cron and opted for a system cron.
What I call "false positive" is a warning shown to say "cron is not running" when it is still not expected to be executed.

Reply:
Yes, I understand. In other words, the plugin requires that either:
- The user enables wp_cron
- Or the user has a system cron that is triggered in a period less than SCHEDULE_OFFSET (30 mins according to the value you set)
Other scenarios are not recommended. If in the cron check calculations we notice that it's taking longer than SCHEDULE_OFFSET, then we will show the warning. This way, we don't need the average calculations, and we can remove the DB queries to the logs. As I said, we should not depend on the logs to do any kind of calculations.